### PR TITLE
adding GoAlert configuration

### PR DESCRIPTION
**Pull request by theautoroboto**
Adding GoAlert to CAMO. This is a FedRAMP requirement to configure GoAlert.

# Changes

- 3 receivers (goalert, goalert-high, goalert-heartbeat) added for GoAlert to replace PagerDuty and DMS services
- 2 routes (goalert and goalert-heartbeat) added for GoAlert to replace PagerDuty and DMS services
- Additional logic added to support GoAlert

### CODE REVIEW

File: controllers/secret_controller.go [hunk 1]
```diff
@@ -220,25 +251,44 @@ func (r *SecretReconciler) SetupWithManager(mgr ctrl.Manager) error {
...
 }
 
-// createPagerdutyRoute creates an AlertManager Route for PagerDuty in memory.
-func createPagerdutyRoute(namespaceList []string) *alertmanager.Route {
+func createSubroutes(reqLogger logr.Logger, namespaceList []string, receiverType string) *alertmanager.Route{
+
+	var receiverCommon, receiverCritical, receiverError, receiverWarning, receiverDefault string
+
+	switch receiverType {
+	case "goalert":
```
Comment (cblecker) [inline controllers/secret_controller.go:259]: Oh one more nit.. can we make "goalert" and "pagerduty" constants, and then replace these strings with that constant wherever they appear in the file?

File: controllers/secret_controller.go [hunk 2]
```diff
@@ -793,17 +906,21 @@ func (r *SecretReconciler) readOCMAgentServiceURLFromConfig(reqLogger logr.Logge
...
-func (r *SecretReconciler) parseSecrets(reqLogger logr.Logger, secretList *corev1.SecretList, namespace string, clusterReady bool) (pagerdutyRoutingKey string, watchdogURL string) {
+func (r *SecretReconciler) parseSecrets(reqLogger logr.Logger, secretList *corev1.SecretList, namespace string, clusterReady bool) (pagerdutyRoutingKey string, watchdogURL string, goalertURLlow string, goalertURLhigh string, goalertURLheartbeat string) {
 	// Check for the presence of specific secrets.
+	goalertSecretExists := secretInList(reqLogger, secretNameGoalert, secretList)
 	pagerDutySecretExists := secretInList(reqLogger, secretNamePD, secretList)
 	snitchSecretExists := secretInList(reqLogger, secretNameDMS, secretList)
 
-	// do the work! collect secret info for PD and DMS
+	// do the work! collect secret info for PD, DMS, and GoAlert
+	goalertURLlow = ""
```
Comment (cblecker) [inline controllers/secret_controller.go:916]: This bit can be changed into the same thing we did on L256

Comment (theautoroboto) [reply controllers/secret_controller.go:916]: These vars are already initialized.  We need them to be blank strings to be populated later in the function based on the criteria

File: controllers/secret_controller.go [hunk 3]
```diff
@@ -220,25 +251,44 @@ func (r *SecretReconciler) SetupWithManager(mgr ctrl.Manager) error {
...
+		receiverWarning = receiverGoAlertLow
+		receiverDefault = receiverGoAlertLow
+	case "pagerduty":
+		receiverCommon = receiverPagerduty
+		receiverCritical = receiverMakeItCritical
+		receiverError = receiverMakeItError
+		receiverWarning = receiverMakeItWarning
+		receiverDefault = defaultReceiver
+	default:
+		reqLogger.Info("INFO: Could not create subroute for receiver", receiverType)
```
[the remark this answers is not in the document]
Comment (theautoroboto) [reply controllers/secret_controller.go:272]: This has been altered to use an enum.  Error detection is now covered by native enum functionality

File: controllers/secret_controller.go [hunk 4]
```diff
@@ -282,15 +332,6 @@ func createPagerdutyRoute(namespaceList []string) *alertmanager.Route {
 		{Receiver: receiverNull, MatchRE: map[string]string{"namespace": alertmanager.PDRegexLP}, Match: map[string]string{"alertname": "TargetDown"}},
 		// https://issues.redhat.com/browse/OSD-5544
 		{Receiver: receiverNull, MatchRE: map[string]string{"job_name": "^elasticsearch.*"}, Match: map[string]string{"alertname": "KubeJobFailed", "namespace": "openshift-logging"}},
-		// https://issues.redhat.com/browse/OSD-11273 - silence all elasticsearch alerts so we can handle only the ones that have extended logging support
```
[the remark this answers is not in the document]
Comment (theautoroboto) [reply controllers/secret_controller.go:285]: Order adjusted to keep the related entries together

File: controllers/secret_controller.go [hunk 5]
```diff
@@ -316,104 +356,106 @@ func createPagerdutyRoute(namespaceList []string) *alertmanager.Route {
...
+		// https://issues.redhat.com/browse/OSD-8006
+		{Receiver: receiverCommon, Match: map[string]string{"alertname": "KubeAPIErrorBudgetBurn", "prometheus": "openshift-monitoring/k8s"}},
 	}
 
 	// Silence insights in FedRAMP until its made available in the environment
 	// https://issues.redhat.com/browse/OSD-13685
 	if config.IsFedramp() {
-		pagerdutySubroutes = append(pagerdutySubroutes,
+		subroute = append(subroute,
 			&alertmanager.Route{Receiver: receiverNull, Match: map[string]string{"alertname": "ClusterOperatorDown", "name": "insights"}},
```
[the remark this answers is not in the document]
Comment (theautoroboto) [reply controllers/secret_controller.go:431]: This was moved to the top

File: controllers/secret_controller.go (hunk 2 above)
Comment (theautoroboto) [reply controllers/secret_controller.go]: There were being declared and initialized as part of the function.  These were not needed and were removed.

File: controllers/secret_controller.go (hunk 1 above)
Comment (theautoroboto) [reply controllers/secret_controller.go]: Fixed by using enum

File: controllers/secret_controller.go [hunk 6]
```diff
@@ -316,104 +356,106 @@ func createPagerdutyRoute(namespaceList []string) *alertmanager.Route {
...
-
 		// https://issues.redhat.com/browse/OSD-14071
 		{Receiver: receiverNull, Match: map[string]string{"alertname": "MultipleDefaultStorageClasses", "namespace": "openshift-cluster-storage-operator"}},
-
 		// https://issues.redhat.com/browse/OSD-14857
 		{Receiver: receiverNull, MatchRE: map[string]string{"mountpoint": "/var/lib/ibmc-s3fs.*"}, Match: map[string]string{"alertname": "NodeFilesystemAlmostOutOfSpace", "severity": "critical"}},
+		// Needed because we are now allowing DMS to continue to allow DMS and GoAlert Heartbeat to coexist. Now we just drop DMS.
+		// {Receiver: receiverNull, Match: map[string]string{"alertname": "SnitchHeartBeat", "severity": "deadman"}},
+		// Needed to drop GoAlert heartbeat alerts
+		{Receiver: receiverNull, Match: map[string]string{"alertname": "Watchdog", "severity": "none"}},
```
[the remark this answers is not in the document]
Comment (theautoroboto) [reply controllers/secret_controller.go]: Moved to the top

**Comment by cblecker** [comment]
/lgtm
/approve
/hold

@theautoroboto When you're ready to test this on stage, you can `/hold cancel`. Thanks for this!

**Comment by theautoroboto** [comment]
@cblecker I ran into a rebase mess.  It's all worked out now.  Can I get another lgtm?

**Pull request merged**